File: examples/rust/quickstart.rs
//! Quickstart: build a state proof and verify it locally.

use lmprk_core::hash::Hash;
use lmprk_core::merkle::Side;
use lmprk_core::{
    hash_leaf, hash_node, MerklePath, MerkleStep, SlotInfo, SlotSnapshot, StateProofBuilder,
};

fn main() {
    let data = b"example-account-state".to_vec();
    let leaf = hash_leaf(&data);
    let sibling = hash_leaf(b"example-sibling");
    let root = hash_node(&leaf, &sibling);

    let snapshot = SlotSnapshot {
        head: SlotInfo {
            slot: 999_000,
            blockhash: Hash([4u8; 32]),
            parent_slot: 998_999,
            signer_count: 110,
        },
        state_root: root,
        validator_set_size: 128,
        threshold: 86,
    };

    let path = MerklePath {
        steps: vec![MerkleStep { sibling, side: Side::Right }],
    };

    let proof = StateProofBuilder::new()
        .snapshot(snapshot)
        .address("9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin")
        .account_data(data)
        .path(path)
        .build()
        .expect("builds");

    proof.verify().expect("verifies");
    println!(
        "proof verified for {} ({} bytes)",
        proof.address,
        proof.byte_size()
    );
}

File: examples/ts/quickstart.ts
import BN from "bn.js";
import { StateProof } from "../../sdk/typescript/src/types";
import { verifyProof } from "../../sdk/typescript/src/verify";

const proof: StateProof = {
  protocol: "lmprk/v1",
  version: 1,
  snapshot: {
    head: {
      slot: new BN(999000),
      blockhash: "00".repeat(32),
      parentSlot: new BN(998999),
      signerCount: 110,
    },
    stateRoot: "00".repeat(32),
    validatorSetSize: 128,
    threshold: 86,
  },
  address: "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin",
  accountData: "deadbeefcafe",
  path: { steps: [] },
};

const result = verifyProof(proof);
if (!result.valid) {
  console.error("proof rejected:", result.reason);
  process.exit(1);
}
console.log("proof accepted, computed root", result.computedRoot);
